src/main_functions.c

```diff
@@ -0,0 +1,530 @@
+		fprintf(f, "%s", lines[i]); // write lines into file
+		i++;
+	}
+	fclose(f);
+}
+
+void save_page(char* filename, char** lines, int* quit)
+{
+	int i = 1;
+    char line[256];
+    char* args = NULL;
+    int verify = 0;
+
+	while (i) {
+        shell_save();
+		shell_prompt();
+		i = 0;
+
+        verify = scanf("%s", line);
+        assert (!(verify < 0)); // ensure valid input
+        assert (strlen(line) < 256); // ensure that string length is within limit
+
+        if (strlen(line) > 2) { // ensure that only one character is entered, otherwise cannot proceed
+            shell_error("Please type in one of the indicated commands!");
+            i = 1;
+        } else if (!strcmp(line, "s")) {
+			save_corrections(filename, lines);  // save to the same file destination, overwriting existing file
+			*quit = 1;
+		} else if (!strcmp(line, "c")) {
+            
+            while ((args == NULL) ^ i) {    // either user returns to main page, or inputs a vaid new destination
+                printf("\n\nEnter a viable file name (*.txt), or enter 'r' to return to the save page.\n\n");
+                shell_prompt();
+                verify = scanf("%s", line);
+                assert (!(verify < 0));
+
+                args = strstr(line, ".txt");
+                if (!strcmp(line, "r")) i = 1;
+            }
+
+            *quit = 1;
+            if (i == 0) save_corrections(line, lines);  // save to different file destination
+
+		} else if (!strcmp(line, "r")) {
+			*quit = 0;
+		} else if (!strcmp(line, "q")) {
+			*quit = 1;
+		} else {
+			shell_error("Please type in one of the indicated commands!");   // wrong input
+			i = 1;
+		}
+
+	}
+}
+
+/*
+ *	II. Functions for editing strings
+ */
+
+void underline_misspelled(char *word, char* underline)
+{
+	int j = strlen(word);
+    int i = 0;
+	for(; i < j; i++) {
+    	strcat(underline, "^");
+	}
+		strcat(underline, " ");
+		// printf("start \n");
+		// printf("%s", underline);
+		// printf("end \n");
+
+}
+
+void underline_correct_spelling(char *word, char* underline)
+{
+	int j = strlen(word);
+    int i = 0;
+	for(; i < j; i++) {
+    		strcat(underline, " ");
+	}
+		strcat(underline, " ");
+
+}
+
+char* underline_misspelled_sentence(char** misspelled, char* sentence, int element) {
+
+	char* underline = malloc(strlen(sentence));
+	underline[0] = '\0'; 
+
+	while(misspelled[element] != NULL) {
+
+		char *ptr = strstr(sentence, misspelled[element]);
+
+		if(ptr != NULL) {
+
+			int pos = ptr - sentence;
+
+			int i = 0;
+			size_t j = 0;
+
+			for(; i < pos; i++) {
+
+				strcat(underline, " ");
+
+			}
+
+			for (; j < strlen(misspelled[element]); j++) {
+
+				strcat(underline, "^");
+
+			}
+
+			sentence = ptr+strlen(misspelled[element]);
+
+		}
+
+		element++;
+
+	}
+
+	return underline;
+
+}
+
+
+
+int add_to_misspelled(char *word, char** misspelled)
+{
+	if (word == NULL || misspelled == NULL) {
+		return EXIT_FAILURE;
+	}
+	int i = 0;
+	while(misspelled[i] != NULL) {
```

Comment:
What happens if we hit the end of misspelled? This will cause a segfault.

Reply:
that won't happen, because misspelled has been initialized to be NULL in initialize_misspelled

Reply:
now NULL in calloc